File: output/anthropic_claude_sonnet_4_6/few/python/Capitais.py
import sys
from collections import deque

def solve():
    input_data = sys.stdin.buffer.read().split()
    idx = 0
    n = int(input_data[idx]); idx += 1
    
    adj = [[] for _ in range(n + 1)]
    degree = [0] * (n + 1)
    
    for _ in range(n - 1):
        u = int(input_data[idx]); idx += 1
        v = int(input_data[idx]); idx += 1
        adj[u].append(v)
        adj[v].append(u)
        degree[u] += 1
        degree[v] += 1
    
    # Capitals are leaves (degree == 1)
    # Non-capitals have degree >= 2
    # We need minimum distance between any pair of capitals
    
    # The path between two capitals goes through internal nodes
    # We need to find the minimum distance between any two leaves
    
    # Key insight: The minimum distance between two leaves is achieved by
    # two leaves that share a common neighbor (internal node), giving distance 2,
    # OR two leaves connected directly (distance 1, only possible if n=2).
    
    # Wait, if n=2, both nodes have degree 1, so both are capitals, distance=1.
    
    # For n>2: can two capitals be adjacent? If u-v is an edge and both have degree 1,
    # that means n=2. For n>2, if u is a leaf and v is a leaf connected to each other,
    # then they have degree 1 but the tree would need to have other nodes connected elsewhere.
    # Actually if u-v edge exists and degree[u]=1 and degree[v]=1, then n=2.
    
    # So for n>2, minimum distance between two leaves:
    # - If an internal node has >= 2 leaf neighbors, answer is 2
    # - Otherwise, we need BFS/tree DP to find minimum distance between leaves
    
    # Check if any internal node has 2+ leaf children -> answer = 2
    leaf_neighbor_count = [0] * (n + 1)
    for v in range(1, n + 1):
        if degree[v] == 1:
            # v is a leaf, its single neighbor
            u = adj[v][0]
            leaf_neighbor_count[u] += 1
    
    for v in range(1, n + 1):
        if leaf_neighbor_count[v] >= 2:
            print(2)
            return
    
    if n == 2:
        print(1)
        return
    
    # No internal node has 2+ leaf neighbors
    # Each internal node has at most 1 leaf neighbor
    # We need minimum distance between two leaves
    # This is a tree problem: find min distance between any two leaves
    
    # Build a compressed tree or do BFS from each leaf? Too slow for n=1e5
    
    # Better approach: For each internal node, track the two closest leaves
    # Use tree DP
    
    # dp[v] = minimum distance from v to a leaf in its subtree (rooted tree)
    # For each node, combine distances from different subtrees
    
    INF = float('inf')
    
    # Root at node with degree >= 2 (any internal node)
    root = -1
    for v in range(1, n + 1):
        if degree[v] >= 2:
            root = v
            break
    
    # Iterative DFS
    parent = [-1] * (n + 1)
    order = []
    visited = [False] * (n + 1)
    stack = [root]
    visited[root] = True
    while stack:
        v = stack.pop()
        order.append(v)
        for u in adj[v]:
            if not visited[u]:
                visited[u] = True
                parent[u] = v
                stack.append(u)
    
    # dp[v] = min distance from v to any leaf in subtree of v
    dp = [INF] * (n + 1)
    
    # Leaves have dp = 0
    for v in range(1, n + 1):
        if degree[v] == 1:
            dp[v] = 0
    
    ans = INF
    
    # Process in reverse order (bottom-up)
    for v in reversed(order):
        # Collect dp values from children
        # For combining: find two smallest dp values from different children
        best1 = INF
        best2 = INF
        
        if degree[v] == 1:
            dp[v] = 0
            continue
        
        for u in adj[v]:
            if u == parent[v]:
                continue
            val = dp[u] + 1
            if val < best1:
                best2 = best1
                best1 = val
            elif val < best2:
                best2 = val
        
        # Also consider parent direction later, but for now:
        if best1 != INF and best2 != INF:
            ans = min(ans, best1 + best2)
        
        dp[v] = best1 if best1 != INF else INF
    
    # Now we need to also consider paths that go through parent
    # Re-root DP: for each node v, also know best distance to leaf going "up"
    
    up = [INF] * (n + 1)
    # up[root] = INF (no parent)
    
    for v in order:
        # Compute up values for children of v
        children_vals = []
        for u in adj[v]:
            if u == parent[v]:
                continue
            children_vals.append((dp[u] + 1, u))
        
        # Also include up[v]+1 as a direction
        all_vals = children_vals[:]
        if up[v] != INF:
            all_vals.append((up[v] + 1, -1))
        
        all_vals.sort()
        
        for u in adj[v]:
            if u == parent[v]:
                continue
            # up[u] = min distance to leaf not going through u
            # = min of (dp[other_child]+1 for other children) and (up[v]+1)
            child_val = dp[u] + 1
            # find best among all_vals excluding u
            best = INF
            for val, node in all_vals:
                if node != u:
                    best = val
                    break
            up[u] = best
            
            # Update answer: path from leaf in u's subtree to leaf via up
            if dp[u] != INF and up[u] != INF:
                ans = min(ans, dp[u] + 1 + up[u])
    
    print(ans)

solve()

File: output/anthropic_claude_sonnet_4_6/few/cpp/Capitais.cpp
#include <bits/stdc++.h>
using namespace std;

int main(){
    ios::sync_with_stdio(0);
    cin.tie(0);
    
    int n;
    cin >> n;
    
    vector<vector<int>> adj(n+1);
    vector<int> deg(n+1, 0);
    
    for(int i = 0; i < n-1; i++){
        int u, v;
        cin >> u >> v;
        adj[u].push_back(v);
        adj[v].push_back(u);
        deg[u]++;
        deg[v]++;
    }
    
    // Capitals are leaves (degree == 1)
    // Non-capitals have degree >= 2
    // We need minimum distance between any pair of capitals
    // Since no path between two capitals passes through another capital,
    // the path between two capitals goes through their non-capital neighbors.
    
    // The minimum distance between two capitals:
    // Each capital is a leaf. The path between two leaves goes through internal nodes.
    // We want the minimum over all pairs of leaves of their distance.
    
    // Key insight: The minimum distance between two leaves is achieved by
    // two leaves that share the same internal neighbor (distance 2),
    // OR if no internal node has degree >= 3 (i.e., it's a path), 
    // then we need to find adjacent leaves or leaves 2 apart.
    
    // Actually: if any internal node has degree >= 3, then there are at least 2 leaf neighbors
    // at distance 2. But internal nodes might not have direct leaf neighbors.
    
    // Better approach: BFS/DFS from each leaf is too slow for N=1e5.
    // 
    // The minimum distance between two leaves in a tree:
    // - If two leaves share the same parent (internal node), distance = 2
    // - Otherwise we need to find the minimum path
    //
    // The minimum distance between any two leaves can be found by:
    // For each internal node, if it has >= 2 leaf children, answer candidate = 2
    // Otherwise, we need to consider paths through the tree.
    //
    // Actually the answer is at least 2 (since leaves can't be adjacent to each other
    // because that would mean both have degree 1 and N=2 special case).
    // Wait, N=2: both nodes have degree 1, they are both capitals, distance=1.
    
    if(n == 2){
        cout << 1 << endl;
        return 0;
    }
    
    // For each internal node, count how many leaf neighbors it has.
    // If any internal node has >= 2 leaf neighbors, answer = 2.
    // Otherwise, we need to find minimum distance.
    // 
    // If no internal node has 2 leaf neighbors, then each internal node has at most 1 leaf neighbor.
    // The tree of internal nodes forms a subtree. We need min distance between two leaves
    // which equals 2 + (distance between their internal neighbors in the internal subtree).
    // Min distance in internal subtree between two nodes that each have a leaf attached = 
    // we need the two closest such nodes in the internal tree.
    // That minimum distance >= 1, so answer >= 3.
    // 
    // To find min distance between internal nodes that have leaf neighbors:
    // BFS/DFS on the internal subtree.
    
    // Check if any internal node has >= 2 leaf neighbors
    for(int v = 1; v <= n; v++){
        if(deg[v] >= 2){ // internal node
            int leafCount = 0;
            for(int u : adj[v]){
                if(deg[u] == 1) leafCount++;
            }
            if(leafCount >= 2){
                cout << 2 << endl;
                return 0;
            }
        }
    }
    
    // No internal node has 2 leaf neighbors.
    // Find minimum distance between internal nodes that have exactly 1 leaf neighbor.
    // BFS from all such nodes simultaneously (multi-source BFS) on internal subgraph.
    
    vector<int> dist(n+1, -1);
    deque<int> q;
    
    // Sources: internal nodes with a leaf neighbor
    for(int v = 1; v <= n; v++){
        if(deg[v] >= 2){
            for(int u : adj[v]){
                if(deg[u] == 1){
                    dist[v] = 0;
                    q.push_back(v);
                    break;
                }
            }
        }
    }
    
    int ans = INT_MAX;
    
    while(!q.empty()){
        int v = q.front(); q.pop_front();
        for(int u : adj[v]){
            if(deg[u] == 1) continue; // skip leaves
            if(dist[u] == -1){
                dist[u] = dist[v] + 1;
                q.push_back(u);
            } else if(dist[u] != dist[v]){ // different source
                // but we can't distinguish sources easily
                // Instead: when we meet a node already visited with different dist
                // Actually multi-source BFS: when we find a node already visited,
                // the distance between two sources through this node is dist[v]+dist[u]+1
                ans = min(ans, dist[v] + dist[u] + 1);
            }
        }
    }
    
    // ans is distance between internal nodes, actual distance between capitals = ans + 2
    if(ans == INT_MAX){
        // Only one internal node has leaf neighbor? Shouldn't happen if n>=3 and tree valid
        cout << -1 << endl;
    } else {
        cout << ans + 2 << endl;
    }
    
    return 0;
}